lib/grandpa/commits_tracker.go

```diff
@@ -90,6 +98,9 @@ func (ct *commitsTracker) delete(blockHash common.Hash) {
 // does not exist in the tracker
 func (ct *commitsTracker) message(blockHash common.Hash) (
 	message *CommitMessage) {
+	ct.Lock()
+	defer ct.Unlock()
```

Comment:
I guess you should lock the `forEach` method as well?

Reply:
It is not possible to lock the `forEach`, the reason is: we call the `forEach` method and we pass to it, as an argument, a function that calls `t.commits.delete` which locks to remove the commit

```go
-> cannot lock for reading
t.commits.forEach(func(...) {
   ...
  -> already locks for delete
   t.commits.delete
})
```

Reply:
But I agree that we should lock while we are getting the commits from the tracker, but with the `forEach` method is not possible so I would like to suggest to replace the `forEach(func(cm *CommitMessage))` for a method that only retrieves the commits in the tracker instead of a method that receives a function and apply the function to each commit in the tracker.

Reply:
Good point, my bad for not checking.

Although, thinking more on this, shouldn't the entire 'forEach' operation lock then? Even if we don't need that much the mutex, one might later assume the commits tracker is thread safe due to most operations locking, so that's a risk for future changes.

> for a method that only retrieves the commits in the tracker

I wrote it as a forEach method for performance if we have a lot of tracked messages. See https://github.com/ChainSafe/gossamer/pull/2490#discussion_r856153149
Maybe there is no need for it though, I haven't had the time to add Prometheus metrics to check this out.

A (hopefully better) alternative would be to add a `deleteWithoutLock` method not locking the mutex, which you can then call in that `forEach` argument function, as well as in the `delete` method.